File: src/minimum-tree-depth.go
package algo

import (
	"math"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func minDepth(root *TreeNode) int {
	min := math.MaxInt

	var dfs func(node *TreeNode, currentDepth int)
	dfs = func(node *TreeNode, currentDepth int) {
		if node.Left == nil && node.Right == nil {
			if currentDepth < min {
				min = currentDepth
			}
		}

		if node.Left != nil {
			dfs(node.Left, currentDepth+1)
		}

		if node.Right != nil {
			dfs(node.Right, currentDepth+1)
		}
	}

	if root == nil {
		return 0
	} else {
		dfs(root, 1)
		return min
	}
}

File: src/minimum-tree-depth.ts
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null

    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

function minDepth(root: TreeNode | null): number {
    let min = Infinity;

    function dfs(node: TreeNode, currentDepth: number) {
        if (!node.left && !node.right) {
            min = Math.min(min, currentDepth);
        }

        if (node.left) {
            dfs(node.left, currentDepth + 1)
        }

        if (node.right) {
            dfs(node.right, currentDepth + 1)
        }
    }


    if (root) {
        dfs(root, 1)
        return min
    } else {
        return 0
    }
}